sound/soc/sof/intel/hda-stream.c

```diff
@@ -440,63 +484,40 @@ irqreturn_t hda_dsp_stream_interrupt(int irq, void *context)
 irqreturn_t hda_dsp_stream_threaded_handler(int irq, void *context)
 {
 	struct snd_sof_dev *sdev = (struct snd_sof_dev *)context;
-	struct sof_intel_hda_dev *hdev = sdev->hda;
+	struct hdac_bus *bus = sof_to_bus(sdev);
+	struct sof_intel_hda_stream *s;
+	//struct sof_intel_hda_dev *hdev = sdev->hda;
 	u32 status = snd_sof_dsp_read(sdev, HDA_DSP_HDA_BAR, SOF_HDA_INTSTS);
 	u32 sd_status;
-	int i;
 
-	/* check playback streams */
-	for (i = 0; i < hdev->num_playback; i++) {
-		/* is IRQ for this stream ? */
-		if (status & (1 << hdev->pstream[i].index)) {
+	/* check streams */
+	list_for_each_entry(s, &bus->stream_list, list) {
+		if (status & (1 << s->index)
+			&& !s->open) {
 			sd_status =
 				snd_sof_dsp_read(sdev, HDA_DSP_HDA_BAR,
-						 hdev->pstream[i].sd_offset +
+						 s->sd_offset +
 						 SOF_HDA_ADSP_REG_CL_SD_STS) &
 						 0xff;
 
-			dev_dbg(sdev->dev, "pstream %d status 0x%x\n",
-				i, sd_status);
+			dev_dbg(sdev->dev, "stream %d status 0x%x\n",
+				s->index, sd_status);
 
 			snd_sof_dsp_update_bits(sdev, HDA_DSP_HDA_BAR,
-						hdev->pstream[i].sd_offset +
+						s->sd_offset +
 						SOF_HDA_ADSP_REG_CL_SD_STS,
 						SOF_HDA_CL_DMA_SD_INT_MASK,
 						SOF_HDA_CL_DMA_SD_INT_MASK);
 
-			if (!hdev->pstream[i].substream ||
-			    !hdev->pstream[i].running ||
+			if (!s->substream ||
+			    !s->running ||
 			    (sd_status & SOF_HDA_CL_DMA_SD_INT_MASK) == 0)
 				continue;
-		}
-	}
-
-	/* check capture streams */
-	for (i = 0; i < hdev->num_capture; i++) {
-		/* is IRQ for this stream ? */
-		if (status & (1 << hdev->cstream[i].index)) {
-			sd_status =
-				snd_sof_dsp_read(sdev, HDA_DSP_HDA_BAR,
-						 hdev->cstream[i].sd_offset +
-						 SOF_HDA_ADSP_REG_CL_SD_STS) &
-						 0xff;
-
-			dev_dbg(sdev->dev, "cstream %d status 0x%x\n",
-				i, sd_status);
-
-			snd_sof_dsp_update_bits(sdev, HDA_DSP_HDA_BAR,
-						hdev->cstream[i].sd_offset +
-						SOF_HDA_ADSP_REG_CL_SD_STS,
-						SOF_HDA_CL_DMA_SD_INT_MASK,
-						SOF_HDA_CL_DMA_SD_INT_MASK);
 
-			if (!hdev->cstream[i].substream ||
-			    !hdev->cstream[i].running ||
-			    (sd_status & SOF_HDA_CL_DMA_SD_INT_MASK) == 0)
-				continue;
 		}
 	}
 
+
```

Comment:
Do we need period_elapsed() ?

Reply:
IMO, no, we do that in IPC interrupt handler.